libbeat/common/transport/tlscommon/tls_config.go

```diff
@@ -116,3 +121,82 @@ func (c *TLSConfig) BuildModuleConfig(host string) *tls.Config {
 	config.ServerName = host
 	return config
 }
+
+// MakeVerifyPeerCertificate creates the verification combination of checking certificate pins and skipping host name validation depending on the config
+func MakeVerifyPeerCertificate(cfg *TLSConfig) verifyPeerCertFunc {
+	pin := len(cfg.CASha256) > 0
+	skipHostName := cfg.Verification == VerifyCertificate
+
+	if pin && !skipHostName {
+		return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+			return VerifyCAPin(cfg.CASha256, verifiedChains)
+		}
+	}
+
+	if pin && skipHostName {
+		return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+			_, _, err := VerifyCertificateExceptServerName(rawCerts, cfg)
+			if err != nil {
+				return err
+			}
+			return VerifyCAPin(cfg.CASha256, verifiedChains)
+		}
+	}
+
+	if !pin && skipHostName {
+		return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+			_, _, err := VerifyCertificateExceptServerName(rawCerts, cfg)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// VerifyCertificateExceptServerName is a TLS Certificate verification utility method that verifies that the provided
+// certificate chain is valid and is signed by one of the root CAs in the provided tls.Config. It is intended to be
+// as similar as possible to the default verify (go/src/crypto/tls/handshake_client.go:259), but does not verify
+// that the provided certificate matches the ServerName in the tls.Config.
+func VerifyCertificateExceptServerName(
+	rawCerts [][]byte,
+	c *TLSConfig,
+) ([]*x509.Certificate, [][]*x509.Certificate, error) {
+	// this is where we're a bit suboptimal, as we have to re-parse the certificates that have been presented
+	// during the handshake.
+	// the verification code here is taken from crypto/tls/handshake_client.go:259
+	certs := make([]*x509.Certificate, len(rawCerts))
+	for i, asn1Data := range rawCerts {
+		cert, err := x509.ParseCertificate(asn1Data)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "tls: failed to parse certificate from server")
```

Comment:
If the goal is to avoid stack traces here then this will also have them, can change it to use %w wrapping if that's preferable

Reply:
yeah, `%w` is definitely prefferable.